reticulate::use_python(
  python = "/usr/bin/python3.8",
  required = TRUE
)
library(reticulate)
py_set_seed(42, disable_hash_randomization = TRUE)
suppressMessages(library(data.table))
suppressMessages(library(dplyr))
suppressMessages(library(tibble))
suppressMessages(library(Matrix))
suppressMessages(library(Seurat))
suppressMessages(library(ggsci))
suppressMessages(library(qs))
set.seed(42)
options(future.globals.maxSize= 100000*1024^3)


library(rstatix)
library(velocyto.R)
library(scales)

#read Seurat object of basophil-lineage clusters in Figure.2 (named as baso_lin)
load("MiyakeWTA2 steady baso single cell.rda") 

#load concatenated loom data generated by velocyto.py
ldat = read.loom.matrices("MiyakeWTA2_velocyto_combined.loom")
emat = ldat$spliced
nmat = ldat$unspliced

JS_res = baso_lin@reductions$pca@jackstraw$overall.p.values
tmp = JS_res[JS_res[,2]>0.05,,drop=F]
if(nrow(tmp)>0){
  dims =c(1:(min(tmp[,1])-1))
} else {
  dims = 1:100
}

#change column nabe (cell ID) for exact match with original Seurat object
hoge = colnames(emat)
hoge = gsub(".*:", "", hoge)
hoge = gsub(".bam", "", hoge)

colnames(emat)=hoge
colnames(nmat)=hoge


#extract required data from Seurat object
dfobs = data.frame(baso_lin@meta.data$RNA_snn_res.0.2, baso_lin@meta.data$RNA_snn_res.0.2)  #specify cluster identity column.
colnames(dfobs)=c("clusters", "celltype")
rownames(dfobs) = gsub("mST07_", "", rownames(baso_lin@meta.data)) #cell names become equal between loom file matrix and Seurat object

dfvar = data.frame(rownames(baso_lin@assays$RNA@counts))
rownames(dfvar) = dfvar[,1]

emb = baso_lin@reductions$umap@cell.embeddings  #extract 2D dimreduction space info. 
pcs = baso_lin@reductions$pca@cell.embeddings
rownames(emb) = gsub("mST07_", "", rownames(emb))
rownames(pcs) = gsub("mST07_", "", rownames(pcs))

#filter data by cell of interest
cell_IDs = intersect(colnames(emat), rownames(dfobs))
gene_IDs = intersect(rownames(emat), rownames(baso_lin@assays$RNA@data))


emat = emat[gene_IDs,cell_IDs]
nmat = nmat[gene_IDs,cell_IDs]
dfobs = dfobs[cell_IDs,]
emb = emb[cell_IDs,]
pcs = pcs[cell_IDs,dims]


dfvar <- data.frame(gene_IDs)
rownames(dfvar) <- gene_IDs
expression_matrix = baso_lin@assays$RNA@data
colnames(expression_matrix)=gsub("mST07_", "", colnames(expression_matrix))
expression_matrix=expression_matrix[gene_IDs, cell_IDs]
expression_matrix=as.matrix(expression_matrix)

#create anndata
#make AnnData object
ad <- import("anndata", convert = FALSE)
scv = import("scvelo", convert = FALSE)


palette = hue_pal()(length(unique(baso_lin@active.ident))) #Seurat default color palette
names(palette)=0:(length(unique(baso_lin@active.ident))-1) #colors named by cluster names

adata <- ad$AnnData(
  X=t(expression_matrix),
  obs=dfobs,
  var=dfvar,
  uns=list('clusters_colors'=palette),
  layers=list('spliced'=t(emat), 'unspliced'=t(nmat)),
  obsm=list('X_UMAP'=emb, 'X_pca'=pcs)   #if you use UMAP, named as X_UMAP instead
)


#check cluster color and plotting
#scv$pl$scatter(adata, legend_loc='lower left', size=60, basis="FItSNE")

## run scvelo dynamic model
scv$pp$filter_and_normalize(adata, min_shared_counts=as.integer(30), min_shared_cells=5) ## filter genes
scv$pp$moments(adata, n_pcs=as.integer(max(dims)), n_neighbors=as.integer(30)) ## normalize and compute moments by using PCs same as Seurat analysis
scv$tl$recover_dynamics(adata, n_jobs=as.integer(24)) ## model

scv$tl$velocity(adata, mode='dynamical')
scv$tl$velocity_graph(adata)

#draw velocity stream vector on UMAP space same as Seurat analysis
scv$pl$velocity_embedding_stream(adata, basis='UMAP',
                                 size = 15, dpi=300,save = "velocity_stream.png", alpha=0.9)


scv$pl$velocity_embedding(adata, basis="UMAP", arrow_length=3, arrow_size=2, dpi=120)

#calculate latent time
scv$tl$latent_time(adata)
scv$pl$scatter(adata, color='latent_time', color_map='gnuplot', size=80, basis='UMAP', save = "latentTime.png")

